pyro/infer/abstract_infer.py

```diff
@@ -317,8 +333,12 @@ def _adjust_to_data(self, trace, data_trace):
                     # Select random sub-indices to replay values under conditionally independent stacks.
                     # Otherwise, we assume there is an dependence of indexes between training data
                     # and prediction data.
-                    subidxs = Categorical(logits=site["value"].new_ones(site["value"].size(cis.dim))).sample([cis.size])
-                    site["value"] = site["value"].index_select(cis.dim, subidxs)
+                    bdim = cis.dim - site["fn"].event_dim
```

Comment:
nit: Can we use more explicit names like `batch_dim` or `cond_indep_dim`? That will make it easier to grok what is happening here. 

Also, I am not sure why we need to adjust by the `event_dim`. I think subsampling from `cis.dim` in the sample site should be enough? Let us add a test case, if that's not the case.

Reply:
I explicitly named the variable now.

The reason that I adjust by event_dim is for distributions like `MultiVariateNormal` or `OneHotCategorical` which have more complex event shapes than scalar outputs from other distributions. I will think of a test to add for this.

Reply:
That makes sense. I think for the test we can just change our existing model/guide from a uniform-binomial to a dirichlet-categorical instead which will exercise this code path. We can handle this in a separate PR, I'll be happy to help with this.